File: Practica_6/Biblioteca/Biblioteca.java
package Practica_6.Biblioteca;

import java.io.Serializable;
import java.awt.Color;
import java.awt.EventQueue;
import java.awt.Font;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.EOFException;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.ArrayList;

import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JTextArea;
import javax.swing.border.EmptyBorder;
public class Biblioteca extends JFrame implements Serializable {

    private static final long serialVersionUID = 1L;
    private JPanel contentPane;
    private String nombre;
    private ArrayList<Libro> libros;
    private ArrayList<Autor> autores;
    private ArrayList<Prestamo> prestamos;
    private Horario horario;
    private JTextArea textArea;

    /**
     * Launch the application.
     */
    public static void main(String[] args) {
        EventQueue.invokeLater(new Runnable() {
            public void run() {
                try {
                    Biblioteca frame = new Biblioteca();
                    frame.setVisible(true);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
    }

    /**
     * Create the frame.
     */
    public Biblioteca() {
        Biblioteca loaded = cargar();
        if (loaded != null) {
            this.nombre = loaded.nombre;
            this.libros = loaded.libros;
            this.autores = loaded.autores;
            this.prestamos = loaded.prestamos;
            this.horario = loaded.horario;
            JOptionPane.showMessageDialog(null, "Biblioteca cargada desde archivo.");
        } else {
            this.nombre = "Biblioteca UMSA";
            this.libros = new ArrayList<>();
            this.autores = new ArrayList<>();
            this.prestamos = new ArrayList<>();
            this.horario = new Horario("Lunes a Viernes", "08:00", "18:00");
            JOptionPane.showMessageDialog(null, "No se encontró archivo. Creando nueva biblioteca.");
        }

        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setBounds(100, 100, 500, 400);
        contentPane = new JPanel();
        contentPane.setBorder(new EmptyBorder(5, 5, 5, 5));
        setContentPane(contentPane);
        contentPane.setLayout(null);

        JPanel panelHeader = new JPanel();
        panelHeader.setBackground(new Color(173, 216, 230));
        panelHeader.setBounds(0, 0, 484, 50);
        contentPane.add(panelHeader);
        panelHeader.setLayout(null);

        JLabel lblTitulo = new JLabel("Gestión de Biblioteca");
        lblTitulo.setFont(new Font("Times New Roman", Font.PLAIN, 20));
        lblTitulo.setBounds(150, 10, 200, 30);
        panelHeader.add(lblTitulo);
        
                JLabel lblImagen = new JLabel("");
                lblImagen.setBounds(284, 11, 200, 50);
                panelHeader.add(lblImagen);
                lblImagen.setIcon(new ImageIcon("C:\\Users\\ASUS\\Desktop\\ALAN 2025\\IMAGENES\\Imagenes_para_informatica\\logo_biblioteca_1.png"));

        JPanel panelBotones = new JPanel();
        panelBotones.setBackground(new Color(240, 240, 240));
        panelBotones.setBounds(0, 50, 150, 311);
        contentPane.add(panelBotones);
        panelBotones.setLayout(null);

        JButton btnAgregarLibro = new JButton("Agregar Libro");
        btnAgregarLibro.setBounds(10, 10, 120, 25);
        btnAgregarLibro.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                agregarLibroUI();
            }
        });
        panelBotones.add(btnAgregarLibro);

        JButton btnAgregarAutor = new JButton("Agregar Autor");
        btnAgregarAutor.setBounds(10, 45, 120, 25);
        btnAgregarAutor.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                agregarAutorUI();
            }
        });
        panelBotones.add(btnAgregarAutor);

        JButton btnPrestarLibro = new JButton("Prestar Libro");
        btnPrestarLibro.setBounds(10, 80, 120, 25);
        btnPrestarLibro.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                prestarLibroUI();
            }
        });
        panelBotones.add(btnPrestarLibro);

        JButton btnMostrarEstado = new JButton("Mostrar Estado");
        btnMostrarEstado.setBounds(10, 115, 120, 25);
        btnMostrarEstado.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                mostrarEstadoUI();
            }
        });
        panelBotones.add(btnMostrarEstado);

        JButton btnGuardar = new JButton("Guardar");
        btnGuardar.setBounds(10, 150, 120, 25);
        btnGuardar.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                guardar();
                JOptionPane.showMessageDialog(null, "Biblioteca guardada.");
            }
        });
        panelBotones.add(btnGuardar);

        JButton btnCargar = new JButton("Cargar");
        btnCargar.setBounds(10, 185, 120, 25);
        btnCargar.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                Biblioteca loaded = cargar();
                if (loaded != null) {
                    nombre = loaded.nombre;
                    libros = loaded.libros;
                    autores = loaded.autores;
                    prestamos = loaded.prestamos;
                    horario = loaded.horario;
                    JOptionPane.showMessageDialog(null, "Biblioteca cargada.");
                    mostrarEstadoUI();
                } else {
                    JOptionPane.showMessageDialog(null, "No se pudo cargar.");
                }
            }
        });
        panelBotones.add(btnCargar);

        JButton btnCerrarBiblio = new JButton("Cerrar Biblio");
        btnCerrarBiblio.setBounds(10, 220, 120, 25);
        btnCerrarBiblio.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                cerrarBiblioteca();
                JOptionPane.showMessageDialog(null, "Préstamos eliminados.");
                mostrarEstadoUI();
            }
        });
        panelBotones.add(btnCerrarBiblio);

        JScrollPane scrollPane = new JScrollPane();
        scrollPane.setBounds(150, 50, 334, 250);
        contentPane.add(scrollPane);

        textArea = new JTextArea();
        scrollPane.setViewportView(textArea);
        textArea.setEditable(false);
    }

    private void agregarLibroUI() {
        String titulo = JOptionPane.showInputDialog("Título del libro:");
        if (titulo == null) return;
        String isbn = JOptionPane.showInputDialog("ISBN:");
        if (isbn == null) return;
        String numPagStr = JOptionPane.showInputDialog("Número de páginas:");
        if (numPagStr == null) return;
        int numPag = Integer.parseInt(numPagStr);
        String[] contenidos = new String[numPag];
        for (int i = 0; i < numPag; i++) {
            contenidos[i] = "Contenido página " + (i + 1);
        }
        Libro libro = new Libro(titulo, isbn, contenidos);
        libros.add(libro);
        JOptionPane.showMessageDialog(null, "Libro agregado.");
    }

    private void agregarAutorUI() {
        String nombre = JOptionPane.showInputDialog("Nombre del autor:");
        if (nombre == null) return;
        String nacionalidad = JOptionPane.showInputDialog("Nacionalidad:");
        if (nacionalidad == null) return;
        Autor autor = new Autor(nombre, nacionalidad);
        autores.add(autor);
        JOptionPane.showMessageDialog(null, "Autor agregado.");
    }

    private void prestarLibroUI() {
        String codEst = JOptionPane.showInputDialog("Código del estudiante:");
        if (codEst == null) return;
        String nomEst = JOptionPane.showInputDialog("Nombre del estudiante:");
        if (nomEst == null) return;
        Estudiante est = new Estudiante(codEst, nomEst);
        String titLib = JOptionPane.showInputDialog("Título del libro:");
        if (titLib == null) return;
        Libro libro = null;
        for (Libro l : libros) {
            if (l.getTitulo().equals(titLib)) {
                libro = l;
                break;
            }
        }
        if (libro != null) {
            String fPrestamo = "2025-11-10";
            String fDevolucion = "2025-11-20";
            Prestamo prestamo = new Prestamo(est, libro, fPrestamo, fDevolucion);
            prestamos.add(prestamo);
            JOptionPane.showMessageDialog(null, "Préstamo creado.");
        } else {
            JOptionPane.showMessageDialog(null, "Libro no encontrado.");
        }
    }

    private void mostrarEstadoUI() {
        String res = "Estado de la Biblioteca " + nombre + ":\n";
        res += horario.mostrarHorario() + "\n";
        res += "Libros disponibles:\n";
        for (Libro l : libros) {
            res += "- " + l.getTitulo() + "\n";
        }
        res += "Autores registrados:\n";
        for (Autor a : autores) {
            res += "- " + a.getNombre() + "\n";
        }
        res += "Préstamos activos:\n";
        for (Prestamo p : prestamos) {
            res += p.mostrarInfo() + "\n";
        }
        textArea.setText(res);
    }

    private void cerrarBiblioteca() {
        prestamos.clear();
    }

    public void guardar() {
        try (FileOutputStream file = new FileOutputStream("biblioteca.dat");
             ObjectOutputStream sal = new ObjectOutputStream(file)) {
            sal.writeObject(this);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static Biblioteca cargar() {
        try (FileInputStream file = new FileInputStream("biblioteca.dat");
             ObjectInputStream ent = new ObjectInputStream(file)) {
            return (Biblioteca) ent.readObject();
        } catch (IOException | ClassNotFoundException e) {
            return null;
        }
    }
}

class Pagina implements Serializable {
    private static final long serialVersionUID = 1L;
    private int numero;
    private String contenido;

    public Pagina(int numero, String contenido) {
        this.numero = numero;
        this.contenido = contenido;
    }

    public String mostrarPagina() {
        return "Página " + numero + ": " + contenido;
    }
}

class Horario implements Serializable {
    private static final long serialVersionUID = 1L;
    private String diasApertura;
    private String horaApertura;
    private String horaCierre;

    public Horario(String dias, String hApertura, String hCierre) {
        this.diasApertura = dias;
        this.horaApertura = hApertura;
        this.horaCierre = hCierre;
    }

    public String mostrarHorario() {
        return "Horario: " + diasApertura + " de " + horaApertura + " a " + horaCierre;
    }
}

class Libro implements Serializable {
    private static final long serialVersionUID = 1L;
    private String titulo;
    private String isbn;
    private java.util.ArrayList<Pagina> paginas;

    public Libro(String titulo, String isbn, String[] contenidosPaginas) {
        this.titulo = titulo;
        this.isbn = isbn;
        this.paginas = new java.util.ArrayList<>();
        for (int i = 0; i < contenidosPaginas.length; i++) {
            paginas.add(new Pagina(i + 1, contenidosPaginas[i]));
        }
    }

    public String getTitulo() {
        return titulo;
    }

    public String leer() {
        String res = "Leyendo libro: " + titulo + "\n";
        for (Pagina p : paginas) {
            res += p.mostrarPagina() + "\n";
        }
        return res;
    }
}

class Autor implements Serializable {
    private static final long serialVersionUID = 1L;
    private String nombre;
    private String nacionalidad;

    public Autor(String nombre, String nacionalidad) {
        this.nombre = nombre;
        this.nacionalidad = nacionalidad;
    }

    public String mostrarInfo() {
        return "Autor: " + nombre + " (" + nacionalidad + ")";
    }

    public String getNombre() {
        return nombre;
    }
}

class Estudiante implements Serializable {
    private static final long serialVersionUID = 1L;
    private String codigo;
    private String nombre;

    public Estudiante(String codigo, String nombre) {
        this.codigo = codigo;
        this.nombre = nombre;
    }

    public String mostrarInfo() {
        return "Estudiante: " + nombre + " (Código: " + codigo + ")";
    }

    public String getNombre() {
        return nombre;
    }
}

class Prestamo implements Serializable {
    private static final long serialVersionUID = 1L;
    private String fechaPrestamo;
    private String fechaDevolucion;
    private Estudiante estudiante;
    private Libro libro;

    public Prestamo(Estudiante estudiante, Libro libro, String fPrestamo, String fDevolucion) {
        this.estudiante = estudiante;
        this.libro = libro;
        this.fechaPrestamo = fPrestamo;
        this.fechaDevolucion = fDevolucion;
    }

    public String mostrarInfo() {
        return "Préstamo: " + libro.getTitulo() + " a " + estudiante.getNombre() +
               " desde " + fechaPrestamo + " hasta " + fechaDevolucion;
    }
}

File: Practica_6/Biblioteca/Biblioteca.py
import json
from tkinter import messagebox

class Biblioteca:
    def __init__(self, nombre, dias, hora_apertura, hora_cierre):
        self.nombre = nombre
        self.libros = []
        self.autores = []
        self.prestamos = []
        self.horario = Horario(dias, hora_apertura, hora_cierre)

    def agregar_libro(self, libro):
        self.libros.append(libro)

    def agregar_autor(self, autor):
        self.autores.append(autor)

    def prestar_libro(self, estudiante, libro, fecha_prestamo, fecha_devolucion):
        if libro in self.libros:
            prestamo = Prestamo(estudiante, libro, fecha_prestamo, fecha_devolucion)
            self.prestamos.append(prestamo)
            return True
        return False

    def mostrar_estado(self):
        res = f"Estado de la Biblioteca {self.nombre}:\n"
        res += self.horario.mostrar_horario() + "\n"
        res += "Libros disponibles:\n"
        for l in self.libros:
            res += f"- {l.get_titulo()}\n"
        res += "Autores registrados:\n"
        for a in self.autores:
            res += f"- {a.get_nombre()}\n"
        res += "Préstamos activos:\n"
        for p in self.prestamos:
            res += p.mostrar_info() + "\n"
        return res

    def cerrar_biblioteca(self):
        self.prestamos = []

    # Persistencia JSON como en el ejemplo: convert to dict
    def to_dict(self):
        return {
            'nombre': self.nombre,
            'horario': self.horario.to_dict(),
            'libros': [l.to_dict() for l in self.libros],
            'autores': [a.to_dict() for a in self.autores],
            'prestamos': [p.to_dict() for p in self.prestamos]
        }

    @classmethod
    def from_dict(cls, d):
        bib = cls(d['nombre'], '', '', '')  # Temp, set horario after
        bib.horario = Horario.from_dict(d['horario'])
        bib.libros = [Libro.from_dict(ld) for ld in d['libros']]
        bib.autores = [Autor.from_dict(ad) for ad in d['autores']]
        bib.prestamos = [Prestamo.from_dict(pd) for pd in d['prestamos']]
        return bib

    def guardar(self):
        with open('biblioteca.json', 'w') as archivo:
            json.dump(self.to_dict(), archivo, indent=4)  # Como en page 24

    @classmethod
    def cargar(cls):
        try:
            with open('biblioteca.json', 'r') as archivo:
                d = json.load(archivo)
            return cls.from_dict(d)
        except FileNotFoundError:
            return None

        bib.horario = Horario.from_dict(d['horario'])
        bib.libros = [Libro.from_dict(ld) for ld in d['libros']]
        bib.autores = [Autor.from_dict(ad) for ad in d['autores']]
        bib.prestamos = [Prestamo.from_dict(pd) for pd in d['prestamos']]
        return bib

    def guardar(self):
        with open('biblioteca.json', 'w') as archivo:
            json.dump(self.to_dict(), archivo, indent=4)  # Como en page 24

    @classmethod
    def cargar(cls):
        try:
            with open('biblioteca.json', 'r') as archivo:
                d = json.load(archivo)
            return cls.from_dict(d)
        except FileNotFoundError:
            return None

# *clases auxiliares
class Pagina:
    def __init__(self, numero, contenido):
        self.numero = numero
        self.contenido = contenido

    def mostrar_pagina(self):
        return f"Página {self.numero}: {self.contenido}"

    def to_dict(self):
        return {'numero': self.numero, 'contenido': self.contenido}

    @classmethod
    def from_dict(cls, d):
        return cls(d['numero'], d['contenido'])

class Horario:
    def __init__(self, dias, hora_apertura, hora_cierre):
        self.dias_apertura = dias
        self.hora_apertura = hora_apertura
        self.hora_cierre = hora_cierre

    def mostrar_horario(self):
        return f"Horario: {self.dias_apertura} de {self.hora_apertura} a {self.hora_cierre}"

    def to_dict(self):
        return {'dias_apertura': self.dias_apertura, 'hora_apertura': self.hora_apertura, 'hora_cierre': self.hora_cierre}

    @classmethod
    def from_dict(cls, d):
        return cls(d['dias_apertura'], d['hora_apertura'], d['hora_cierre'])

class Libro:
    def __init__(self, titulo, isbn, contenidos_paginas):
        self.titulo = titulo
        self.isbn = isbn
        self.paginas = [Pagina(i+1, cont) for i, cont in enumerate(contenidos_paginas)]

    def get_titulo(self):
        return self.titulo

    def leer(self):
        res = f"Leyendo libro: {self.titulo}\n"
        for p in self.paginas:
            res += p.mostrar_pagina() + "\n"
        return res

    def to_dict(self):
        return {'titulo': self.titulo, 'isbn': self.isbn, 'paginas': [p.to_dict() for p in self.paginas]}

    @classmethod
    def from_dict(cls, d):
        contenidos = [pd['contenido'] for pd in d['paginas']]  # Solo necesitamos contenidos para recrear
        return cls(d['titulo'], d['isbn'], contenidos)

class Autor:
    def __init__(self, nombre, nacionalidad):
        self.nombre = nombre
        self.nacionalidad = nacionalidad

    def mostrar_info(self):
        return f"Autor: {self.nombre} ({self.nacionalidad})"

    def get_nombre(self):
        return self.nombre

    def to_dict(self):
        return {'nombre': self.nombre, 'nacionalidad': self.nacionalidad}

    @classmethod
    def from_dict(cls, d):
        return cls(d['nombre'], d['nacionalidad'])

class Estudiante:
    def __init__(self, codigo, nombre):
        self.codigo = codigo
        self.nombre = nombre

    def mostrar_info(self):
        return f"Estudiante: {self.nombre} (Código: {self.codigo})"

    def get_nombre(self):
        return self.nombre

    def to_dict(self):
        return {'codigo': self.codigo, 'nombre': self.nombre}

    @classmethod
    def from_dict(cls, d):
        return cls(d['codigo'], d['nombre'])

class Prestamo:
    def __init__(self, estudiante, libro, fecha_prestamo, fecha_devolucion):
        self.estudiante = estudiante
        self.libro = libro
        self.fecha_prestamo = fecha_prestamo
        self.fecha_devolucion = fecha_devolucion

    def mostrar_info(self):
        return f"Préstamo: {self.libro.get_titulo()} a {self.estudiante.get_nombre()} desde {self.fecha_prestamo} hasta {self.fecha_devolucion}"

    def to_dict(self):
        return {'estudiante': self.estudiante.to_dict(), 'libro': self.libro.to_dict(),
                'fecha_prestamo': self.fecha_prestamo, 'fecha_devolucion': self.fecha_devolucion}

    @classmethod
    def from_dict(cls, d):
        est = Estudiante.from_dict(d['estudiante'])
        lib = Libro.from_dict(d['libro'])
        return cls(est, lib, d['fecha_prestamo'], d['fecha_devolucion'])

# *Tkinter
import tkinter as tk
from tkinter import simpledialog, Text, Scrollbar

class Menu(tk.Tk):
    def __init__(self):
        super().__init__()
        self.title("MENU")
        self.geometry("450x300")

        # Header azul
        header = tk.Frame(self, bg="blue", height=57)
        header.pack(fill="x")

        lbl = tk.Label(header, text="MENU", fg="white", font=("Times New Roman", 25))
        lbl.place(x=163, y=11)

        # Panel principal
        panel = tk.Frame(self, height=207)
        panel.pack(fill="x")

        btn_biblio = tk.Button(panel, text="Biblioteca", command=self.open_biblioteca)
        btn_biblio.place(x=27, y=22)

        # Otros botones placeholder
        tk.Button(panel, text="Libro").place(x=27, y=60)
        tk.Button(panel, text="Autor").place(x=27, y=104)
        tk.Button(panel, text="Estudiante").place(x=27, y=146)

        # Logo
        try:
            img = tk.PhotoImage(file="C:\\Users\\ASUS\\Desktop\\ALAN 2025\\IMAGENES\\Imagenes_para_informatica\\logo_biblioteca_1.png")
            lbl_img = tk.Label(panel, image=img)
            lbl_img.image = img  # Keep reference
            lbl_img.place(x=208, y=32)
        except:
            pass  # Si no carga, ignora

    def open_biblioteca(self):
        biblio_win = BibliotecaWindow()
        biblio_win.mainloop()

class BibliotecaWindow(tk.Tk):
    def __init__(self):
        super().__init__()
        self.title("Gestión de Biblioteca")
        self.geometry("500x400")

        self.biblio = Biblioteca.cargar()
        if self.biblio is None:
            self.biblio = Biblioteca("Biblioteca UMSA", "Lunes a Viernes", "08:00", "18:00")
            messagebox.showinfo("Info", "No se encontró archivo. Creando nueva.")

        # Header azul claro
        header = tk.Frame(self, bg="#ADD8E6", height=50)
        header.pack(fill="x")

        lbl = tk.Label(header, text="Gestión de Biblioteca", font=("Times New Roman", 20))
        lbl.place(x=150, y=10)

        # Panel botones gris claro
        panel_botones = tk.Frame(self, bg="#F0F0F0", width=150)
        panel_botones.pack(side="left", fill="y")

        tk.Button(panel_botones, text="Agregar Libro", command=self.agregar_libro_ui).pack(pady=5)
        tk.Button(panel_botones, text="Agregar Autor", command=self.agregar_autor_ui).pack(pady=5)
        tk.Button(panel_botones, text="Prestar Libro", command=self.prestar_libro_ui).pack(pady=5)
        tk.Button(panel_botones, text="Mostrar Estado", command=self.mostrar_estado_ui).pack(pady=5)
        tk.Button(panel_botones, text="Guardar", command=self.guardar_ui).pack(pady=5)
        tk.Button(panel_botones, text="Cargar", command=self.cargar_ui).pack(pady=5)
        tk.Button(panel_botones, text="Cerrar Biblio", command=self.cerrar_biblio_ui).pack(pady=5)

        # Text area
        scroll = Scrollbar(self)
        self.text_area = Text(self, wrap="word", yscrollcommand=scroll.set)
        scroll.config(command=self.text_area.yview)
        self.text_area.pack(side="left", fill="both", expand=True)
        scroll.pack(side="right", fill="y")

        # Imagen estantería al pie (asume path)
        try:
            img_shelves = tk.PhotoImage(file="library_shelves.jpg")
            lbl_shelves = tk.Label(self, image=img_shelves)
            lbl_shelves.image = img_shelves
            lbl_shelves.pack(side="bottom")
        except:
            pass

    def agregar_libro_ui(self):
        titulo = simpledialog.askstring("Agregar Libro", "Título:")
        if not titulo: return
        isbn = simpledialog.askstring("Agregar Libro", "ISBN:")
        if not isbn: return
        num_pag = simpledialog.askinteger("Agregar Libro", "Número de páginas:")
        if not num_pag: return
        contenidos = [f"Contenido página {i+1}" for i in range(num_pag)]
        libro = Libro(titulo, isbn, contenidos)
        self.biblio.agregar_libro(libro)
        messagebox.showinfo("Éxito", "Libro agregado.")

    def agregar_autor_ui(self):
        nombre = simpledialog.askstring("Agregar Autor", "Nombre:")
        if not nombre: return
        nacionalidad = simpledialog.askstring("Agregar Autor", "Nacionalidad:")
        if not nacionalidad: return
        autor = Autor(nombre, nacionalidad)
        self.biblio.agregar_autor(autor)
        messagebox.showinfo("Éxito", "Autor agregado.")

    def prestar_libro_ui(self):
        cod_est = simpledialog.askstring("Prestar", "Código estudiante:")
        if not cod_est: return
        nom_est = simpledialog.askstring("Prestar", "Nombre estudiante:")
        if not nom_est: return
        est = Estudiante(cod_est, nom_est)
        tit_lib = simpledialog.askstring("Prestar", "Título libro:")
        if not tit_lib: return
        libro = next((l for l in self.biblio.libros if l.get_titulo() == tit_lib), None)
        if libro and self.biblio.prestar_libro(est, libro, "2025-11-10", "2025-11-20"):
            messagebox.showinfo("Éxito", "Préstamo creado.")
        else:
            messagebox.showerror("Error", "Libro no encontrado.")

    def mostrar_estado_ui(self):
        self.text_area.delete(1.0, tk.END)
        self.text_area.insert(tk.END, self.biblio.mostrar_estado())

    def guardar_ui(self):
        self.biblio.guardar()
        messagebox.showinfo("Éxito", "Guardado en JSON.")

    def cargar_ui(self):
        loaded = Biblioteca.cargar()
        if loaded:
            self.biblio = loaded
            messagebox.showinfo("Éxito", "Cargado.")
            self.mostrar_estado_ui()
        else:
            messagebox.showerror("Error", "No se pudo cargar.")

    def cerrar_biblio_ui(self):
        self.biblio.cerrar_biblioteca()
        messagebox.showinfo("Éxito", "Préstamos eliminados.")
        self.mostrar_estado_ui()

if __name__ == "__main__":
    menu = Menu()
    menu.mainloop()